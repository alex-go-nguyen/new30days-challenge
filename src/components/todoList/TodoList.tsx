import React from 'react';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';
import { yupResolver } from '@hookform/resolvers/yup';
import classNames from 'classnames/bind';
import styles from './TodoList.module.scss';
import { FaTrash } from 'react-icons/fa';
import { v4 as uuidv4 } from 'uuid';
import { useLocalStorage } from '../../hooks/useLocalStorage';

const cx = classNames.bind(styles);

type Item = {
    id: string;
    content: string;
};
export interface TodoListProps {
    data: Item[];
}

type UserSubmitForm = {
    todo: string;
};

const schema = yup.object<UserSubmitForm>({
    todo: yup.string().required(),
});

export default function TodoList({ data = [] }: TodoListProps) {
    const [todos, setTodos] = useLocalStorage<Item[]>('todo', data);
    const {
        register,
        handleSubmit,
        formState: { errors },
        reset,
    } = useForm<UserSubmitForm>({ resolver: yupResolver(schema) });

    const onSubmitHandler = (data: UserSubmitForm) => {
        setTodos((prev) => [...prev, { id: uuidv4(), content: data.todo }]);
        reset();
    };

    const handleRemoveTodo = (id: string) => setTodos((prev) => prev.filter((item) => item.id !== id));

    return (
        <div className={cx('container')}>
            <div className={cx('input-wrapper')}>
                <form onSubmit={handleSubmit(onSubmitHandler)}>
                    <input {...register('todo')} placeholder="Enter your todo" />
                    <button className={cx('btn-add')} type="submit">
                        Add
                    </button>
                </form>
            </div>
            <div className={cx('todo-list')}>
                {todos.map((todo) => (
                    <div className={cx('item')} key={todo.id}>
                        <p>{todo.content}</p>
                        <span className={cx('btn-remove')} onClick={() => handleRemoveTodo(todo.id)}>
                            <FaTrash />
                        </span>
                    </div>
                ))}
            </div>
        </div>
    );
}
