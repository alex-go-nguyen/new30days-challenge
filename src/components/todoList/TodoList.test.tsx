import React from 'react';
import renderer from 'react-test-renderer';
import TodoList from './TodoList';

describe('TodoList component', () => {
    const component = renderer.create(
        <TodoList
            data={[
                { id: '1', content: 'HTML' },
                { id: '2', content: 'CSS' },
            ]}
        />,
    );
    let tree = component.toJSON();
    test('should render correctly', () => {
        expect(tree).toMatchSnapshot();
    });
});
