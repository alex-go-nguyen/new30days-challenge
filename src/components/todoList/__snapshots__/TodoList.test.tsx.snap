// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should render correctly 1`] = `
<div
  className="container"
>
  <div
    className="input-wrapper"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        name="todo"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Enter your todo"
      />
      <button
        className="btn-add"
        type="submit"
      >
        Add
      </button>
    </form>
  </div>
  <div
    className="todo-list"
  >
    <div
      className="item"
    >
      <p>
        HTML
      </p>
      <span
        className="btn-remove"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
          />
        </svg>
      </span>
    </div>
    <div
      className="item"
    >
      <p>
        CSS
      </p>
      <span
        className="btn-remove"
        onClick={[Function]}
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
